velox/functions/prestosql/MapSubset.h

```diff
@@ -21,77 +21,71 @@
-        if (!entry.second.has_value()) {                            \
-          auto& keyWriter = out.add_null();                         \
-          keyWriter = entry.first;                                  \
-        } else {                                                    \
-          auto [keyWriter, valueWriter] = out.add_item();           \
-          keyWriter = entry.first;                                  \
-          valueWriter.copy_from(entry.second.value());              \
-        }                                                           \
-                                                                    \
-        --toFind;                                                   \
-        if (toFind == 0) {                                          \
-          break;                                                    \
-        }                                                           \
-      }                                                             \
-    }                                                               \
-                                                                    \
-   private:                                                         \
-    void initializeSearchKeys(const arg_type<Array<TType>>& keys) { \
-      for (const auto& key : keys.skipNulls()) {                    \
-        searchKeys_.emplace(key);                                   \
-      }                                                             \
-    }                                                               \
-                                                                    \
-    bool constantSearchKeys_{false};                                \
-    folly::F14FastSet<TType> searchKeys_;                           \
-  };
-
-VELOX_MAP_SUBSET_PRIMITIVE_FUNCTION(Integer, int32_t)
-VELOX_MAP_SUBSET_PRIMITIVE_FUNCTION(Bigint, int64_t)
-
-#undef VELOX_MAP_SUBSET_PRIMITIVE_FUNCTION
+template <typename TExec, typename Key>
+struct MapSubsetPrimitiveFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig& /*config*/,
+      const arg_type<Map<Key, Generic<T1>>>* /*inputMap*/,
+      const arg_type<Array<Key>>* keys) {
+    if (keys != nullptr) {
+      constantSearchKeys_ = true;
+      initializeSearchKeys(*keys);
+    }
+  }
+
+  void call(
+      out_type<Map<Key, Generic<T1>>>& out,
+      const arg_type<Map<Key, Generic<T1>>>& inputMap,
+      const arg_type<Array<Key>>& keys) {
+    if (keys.empty()) {
+      return;
+    }
+
+    if (!constantSearchKeys_) {
+      searchKeys_.clear();
+      initializeSearchKeys(keys);
+    }
+
+    if (searchKeys_.empty()) {
+      return;
+    }
+
+    auto toFind = searchKeys_.size();
+
+    for (const auto& entry : inputMap) {
+      if (!searchKeys_.contains(entry.first)) {
+        continue;
+      }
+
+      if (!entry.second.has_value()) {
+        auto& keyWriter = out.add_null();
```

Comment:
when value is Generic we can deal with it as primitive and easily support here
out.emplace(entry.first, std::nullopt);
We can even avoid the whole if else and do 
out.emplace(entry.first, entry.second)
instead of 
```
if (!entry.second.has_value()) {
        auto& keyWriter = out.add_null();
        keyWriter = entry.first;
      } else {
        auto [keyWriter, valueWriter] = out.add_item();
        keyWriter = entry.first;
        valueWriter.copy_from(entry.second.value());
      }
      ```
      I can try to put a quick diff for that, sounds useful!

Reply:
Thanks @laithsakka That will be awesome if you can submit a change, we can use here.